six/modules/c++/six.sicd/source/SCPCOA.cpp

```diff
@@ -59,3 +67,314 @@ bool SCPCOA::operator==(const SCPCOA& rhs) const
         azimAngle == rhs.azimAngle &&
         layoverAngle == rhs.layoverAngle);
 }
+
+double SCPCOA::derivedSCPTime(const Grid& grid) const
+{
+    return grid.timeCOAPoly[0][0];
+}
+
+double SCPCOA::derivedSlantRange(const Vector3& scp) const
+{
+    return (scp - arpPos).norm();
+}
+
+double SCPCOA::derivedGroundRange(const Vector3& scp) const
+{
+    return scp.norm() * arpPos.angle(scp);
+}
+
+double SCPCOA::derivedDopplerConeAngle(const Vector3& scp) const
+{
+    return std::acos(arpVel.unit().dot(uLOS(scp))) *
+            math::Constants::RADIANS_TO_DEGREES;
+}
+
+Vector3 SCPCOA::earthTangentPlane(const Vector3& scp) const
+{
+    return Utilities::wgs84Norm(scp);
+}
+
+double SCPCOA::derivedGrazeAngle(const Vector3& scp) const
+{
+    return std::asin(earthTangentPlane(scp).dot(uLOS(scp) * -1)) *
+            math::Constants::RADIANS_TO_DEGREES;
+}
+
+double SCPCOA::derivedIncidenceAngle(const Vector3& scp) const
+{
+    return 90 - derivedGrazeAngle(scp);
+}
+
+double SCPCOA::derivedSlopeAngle(const Vector3& scp) const
+{
+    return std::acos(earthTangentPlane(scp).dot(slantPlaneNormal(scp))) *
+            math::Constants::RADIANS_TO_DEGREES;
+}
+
+Vector3 SCPCOA::slantPlaneNormal(const Vector3& scp) const
+{
+    return cross(arpVel, uLOS(scp)).unit() * look(scp);
+}
+
+Vector3 SCPCOA::uGPX(const Vector3& scp) const
+{
+    return ((uLOS(scp) * -1) -
+        earthTangentPlane(scp) *
+        (earthTangentPlane(scp).dot(uLOS(scp) * -1))).unit();
+}
+
+Vector3 SCPCOA::uEast(const Vector3& scp) const
+{
+    return cross(uNorth(scp), earthTangentPlane(scp));
+}
+
+Vector3 SCPCOA::uNorth(const Vector3& scp) const
+{
+    std::vector<double> coordinates;
+    coordinates.resize(3);
+    coordinates[0] = 0;
+    coordinates[1] = 0;
+    coordinates[2] = 1;
+    Vector3 base(coordinates);
+
+    // Project north onto ground plane
+    Vector3 northGround = base -
+            (earthTangentPlane(scp) * earthTangentPlane(scp).dot(base));
+    return northGround.unit();
+}
+
+double SCPCOA::derivedTwistAngle(const Vector3& scp) const
+{
+    // Angle from +GPY axis to the +SPY axis in plane of incidence
```

Comment:
For this and pretty much all of these derived angle methods... can you get them from a `SceneGeometry` object?  See `six::sicd::Utilities::getSceneGeometry()` for how to make one.  If it's not obvious how to do this cleanly because you've just got an `SCPCOA` object, let's talk.